### ISSUE TITLE
Determination of the type of an object with uninitialized fields

### ISSUE THREAD

**Issue by d01c2** [tc39/ecma262#3482]
I have a question about **how to determine the type of an object with uninitialized fields**, as a continuation of the discussion in [PR #3383](https://github.com/tc39/ecma262/pull/3383).

Let’s assume the execution process of the program below:
```js
new Promise(function () {});
```

This program follows the step of [27.2.3.1 Promise ( _executor_ )](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-promise-executor) during execution. In step 8, it calls [27.2.1.3 CreateResolvingFunctions( _promise_ )](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-createresolvingfunctions) with an argument `_promise_`, whose `[[PromiseResult]]` field is **uninitialized**. In this example, the annotated parameter type of the [callee](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-createresolvingfunctions) is **a Promise**, and [27.2.6 Properties of Promise Instances](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-properties-of-promise-instances) says the `[[PromiseResult]]` field of a promise instance should hold an ECMAScript language value.

I'm unsure if we can interpret the actual argument above as **a Promise**.

Similarly, the same JavaScript program also follows the steps of [27.5.3.8 CreateIteratorFromClosure ( _closure_, _generatorBrand_, _generatorPrototype_ [ , _extraSlots_ ] )](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-createiteratorfromclosure). It calls[27.5.3.1 GeneratorStart ( _generator_, _generatorBody_ )](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-generatorstart) in step 14. The passed argument `_generator_` has an **uninitialized** `[[GeneratorContext]]` field, which should be initialized in this case.

Additionally, I'm aware of the following statement:

> Unless specified otherwise, the initial value of an internal slot is the value **undefined**.

which is specified in [6.1.7.2 Object Internal Methods and Internal Slots](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-object-internal-methods-and-internal-slots). Does this statement mean that the value of uninitialized field should be considered as `undefined` in case of type determination of an object?

**Comment by michaelficarra** [comment]
> I'm unsure if we can interpret the actual argument above as **a Promise**.

I agree that this is unclear. We're going to try to clear up the initialisation of internal slots for objects created this way as part of https://github.com/tc39/ecma262/issues/3399.

I'll leave this open as a reminder for us to update our description of the various types of objects that are created in this way to define them in terms of their essential internal slots. I think those two changes together should resolve your concerns.

**Comment by bakkot** [comment]
#3491 handles the Promise part of this.
